package com.tianji.promotion.service.impl;

import com.baomidou.mybatisplus.extension.plugins.pagination.Page;
import com.tianji.common.domain.dto.PageDTO;
import com.tianji.common.exceptions.BadRequestException;
import com.tianji.common.exceptions.BizIllegalException;
import com.tianji.common.utils.BeanUtils;
import com.tianji.common.utils.CollUtils;
import com.tianji.common.utils.StringUtils;
import com.tianji.promotion.domain.dto.CouponFormDTO;
import com.tianji.promotion.domain.dto.CouponIssueFormDTO;
import com.tianji.promotion.domain.po.Coupon;
import com.tianji.promotion.domain.po.CouponScope;
import com.tianji.promotion.domain.query.CouponQuery;
import com.tianji.promotion.domain.vo.CouponDetailVO;
import com.tianji.promotion.domain.vo.CouponPageVO;
import com.tianji.promotion.enums.CouponStatus;
import com.tianji.promotion.enums.ObtainType;
import com.tianji.promotion.mapper.CouponMapper;
import com.tianji.promotion.service.ICouponScopeService;
import com.tianji.promotion.service.ICouponService;
import com.baomidou.mybatisplus.extension.service.impl.ServiceImpl;
import com.tianji.promotion.service.IExchangeCodeService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

import static com.tianji.promotion.enums.CouponStatus.*;

/**
 * <p>
 * 优惠券的规则信息 服务实现类
 * </p>
 *
 * @author kaii
 * @since 2026-06-19
 */
@Service
@RequiredArgsConstructor
public class CouponServiceImpl extends ServiceImpl<CouponMapper, Coupon> implements ICouponService {
    private final ICouponScopeService couponScopeService;
    private final IExchangeCodeService codeService;

    /*
    * 新增优惠劵
    * */
    @Override
    public void saveCoupon(CouponFormDTO dto) {
        // 1.保存优惠券
        // 1.1.转PO
        Coupon coupon = BeanUtils.copyBean(dto, Coupon.class);
        // 1.2.保存
        this.save(coupon);

        if (!dto.getSpecific()) {
            // 没有范围限定
            return;
        }
        Long couponId = coupon.getId();
        // 2.保存限定范围
        List<Long> scopes = dto.getScopes();
        if (CollUtils.isEmpty(scopes)) {
            throw new BadRequestException("限定范围不能为空");
        }
        // 2.1.转换PO
        List<CouponScope> list = scopes.stream()
                .map(bizId -> new CouponScope().setBizId(bizId).setCouponId(couponId))
                .collect(Collectors.toList());
        // 2.2.保存
        couponScopeService.saveBatch(list);
    }



    /*
    * 分页查找优惠券
    * */
    @Override
    public PageDTO<CouponPageVO> queryCouponByPage(CouponQuery query) {
        Integer status = query.getStatus();
        String name = query.getName();
        Integer type = query.getType();
        // 1.分页查询
        Page<Coupon> page = lambdaQuery()
                .eq(type != null, Coupon::getDiscountType, type)
                .eq(status != null, Coupon::getStatus, status)
                .like(StringUtils.isNotBlank(name), Coupon::getName, name)
                .page(query.toMpPageDefaultSortByCreateTimeDesc());
        // 2.处理VO
        List<Coupon> records = page.getRecords();
        if (CollUtils.isEmpty(records)) {
            return PageDTO.empty(page);
        }
        List<CouponPageVO> list = BeanUtils.copyList(records, CouponPageVO.class);
        // 3.返回
        return PageDTO.of(page, list);
    }


    /*
    * 发放优惠券
    * */
    @Override
    public void beginIssue(CouponIssueFormDTO dto) {
        // 1.查询优惠券
        Coupon coupon = getById(dto.getId());
        if (coupon == null) {
            throw new BadRequestException("优惠券不存在！");
        }
        // 2.判断优惠券状态，是否是暂停或待发放
        if(coupon.getStatus() != CouponStatus.DRAFT && coupon.getStatus() != PAUSE){
            throw new BizIllegalException("优惠券状态错误！");
        }
        // 3.判断是否是立刻发放
        LocalDateTime issueBeginTime = dto.getIssueBeginTime();
        LocalDateTime now = LocalDateTime.now();
        boolean isBegin = issueBeginTime == null || !issueBeginTime.isAfter(now);
        // 4.更新优惠券
        // 4.1.拷贝属性到PO
        Coupon c = BeanUtils.copyBean(dto, Coupon.class);
        // 4.2.更新状态
        if (isBegin) {
            c.setStatus(ISSUING);
            c.setIssueBeginTime(now);
        }else{
            c.setStatus(UN_ISSUE);
        }
        // 4.3.写入数据库
        this.updateById(c);

        //  兑换码生成
        // 5.判断是否需要生成兑换码，优惠券类型必须是兑换码，优惠券状态必须是待发放
        if(coupon.getObtainWay() == ObtainType.ISSUE && coupon.getStatus() == CouponStatus.DRAFT){
            coupon.setIssueEndTime(c.getIssueEndTime());
            codeService.asyncGenerateCode(coupon);
        }
    }


    /*
    * 根据优惠券ID修改优惠券
    * */
    @Override
    public void updateCouponById(Long id, CouponFormDTO dto) {
        // 校验参数
        Long dtoId = dto.getId();
        if ((dtoId != null && id != null && !dtoId.equals(id)) || (dtoId == null && id == null)) {
            throw new BadRequestException("参数错误");
        }
        // 更新优惠券基本信息
        Coupon coupon = BeanUtils.copyBean(dto, Coupon.class);
        // 只更新状态为1的优惠券基本信息，如果失败则是状态已修改
        boolean update = lambdaUpdate().eq(Coupon::getStatus, 1).update(coupon);
        // 基本信息更新失败则无需更新优惠券范围信息
        if (!update) {
            return;
        }
        //  更新优惠券范围信息
        List<Long> scopeIds = dto.getScopes();
        // 优惠券不满减，或优惠券范围为空，则不更新优惠券范围信息
        // 先删除优惠券范围信息，再重新插入
        List<Long> ids = couponScopeService.lambdaQuery()
                .select(CouponScope::getId).eq(CouponScope::getCouponId, dto.getId()).list()
                .stream().map(CouponScope::getId).collect(Collectors.toList());
        couponScopeService.removeByIds(ids);
        // 删除成功后，并且有范围再插入
        if (CollUtils.isNotEmpty(scopeIds)) {
            List<CouponScope> lis = scopeIds.stream()
                    .map(i -> new CouponScope().setCouponId(dto.getId()).setType(1).setBizId(i))
                    .collect(Collectors.toList());
            couponScopeService.saveBatch(lis);
        }
    }

    /*
    * 根据优惠券ID删除
    * */
    @Override
    public void deleteCoupon(Long id) {
        Coupon coupon = getById(id);
        if(coupon == null){
            throw new BadRequestException("优惠券 id 不能为空!");
        }
        if(coupon.getStatus() != DRAFT){
            throw new BadRequestException("优惠券为非待发放状态，不允许删除");
        }
        boolean remove = removeById(id);
        if(remove){
            //删除关联表数据
            List<CouponScope> list = couponScopeService.lambdaQuery()
                    .eq(CouponScope::getCouponId, id)
                    .list();
            if(CollUtils.isNotEmpty(list)){
                //remove
                List<Long> r = list.stream()
                        .map(CouponScope::getId)
                        .collect(Collectors.toList());
                couponScopeService.removeByIds(r);
            }
        }


    }


    /*
    * 根据ID查询优惠券
    * */
    @Override
    public CouponDetailVO getCoupon(Long id) {
        Coupon coupon = getById(id);
        if(coupon == null){
            throw new BadRequestException("优惠券不存在!");
        }
        CouponDetailVO vo = BeanUtils.copyBean(coupon, CouponDetailVO.class);
        //查询分类
        couponScopeService.lambdaQuery().eq(CouponScope::getCouponId)


        return vo;
    }


}
